### PR TITLE
imageoptim-cli 2.0.2 (new formula)

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

### CODE REVIEW

File: Formula/imageoptim-cli.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      You can install with Homebrew-Cask:
+        brew cask install #{cask} --appdir=/Applications
+    EOS
+  end
+end
+
+class ImageoptimCli < Formula
+  desc "CLI interface for ImageOptim, ImageAlpha and JPEGmini"
+  homepage "https://jamiemason.github.io/ImageOptim-CLI/"
+  url "https://registry.npmjs.org/imageoptim-cli/-/imageoptim-cli-2.0.0.tgz"
```
Comment (commitay) [inline Formula/imageoptim-cli.rb]: `2.0.2`

**Comment by commitay** [comment]
@BrewTestBot test this please

**Comment by commitay** [comment]
Can you change the commit subject to match the PR title please?

**Comment by commitay** [comment]
Not sure how this can be bottled as the requirement won't be satisfied on CI.

**Comment by MikeMcQuaid** [comment]
> Not sure how this can be bottled as the requirement won't be satisfied on CI.

Yeh, I don't think we should be adding formulae that require on non-X11/Java casks for that reason to Homebrew/homebrew-core.

**Comment by ilovezfs** [comment]
does it actually require the cask to build? If not, then the requirement could be removed.

**Comment by reitermarkus** [comment]
Not required to build, just to run.

**Comment by ilovezfs** [comment]
It looks like node isn't actually needed either and this can be bottle unneeded.

**Comment by ilovezfs** [comment]
i.e.
```
  bottle :unneeded

  def install
    libexec.install Dir["*"]
    bin.install_symlink libexec/"dist/imageoptim"
  end

  test do
    assert_match version.to_s, shell_output("#{bin}/imageoptim -V")
  end
```
I don't feel strongly about whether we want to consider it in scope for https://github.com/Homebrew/homebrew-core or a better fit for cask.

**Comment by ilovezfs** [comment]
Actually, it looks like `dist/imageoptim` is a prebuilt Mach-O so we can't do that.

**Comment by ilovezfs** [comment]
Similarly, the npm-based installation currently in the PR is also installing that prebuilt Mach-O, so that's not an option either. We'd need to build `imageoptim` from source for this to be in https://github.com/Homebrew/homebrew-core

**Comment by reitermarkus** [comment]
It's now properly building from source, also removed the requirement.

File: Formula/imageoptim-cli.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  homepage "https://jamiemason.github.io/ImageOptim-CLI/"
+  url "https://github.com/JamieMason/ImageOptim-CLI/archive/2.0.2.tar.gz"
+  sha256 "65ccf26cf45012e20877bf7985b1f9d99cfef4ac3d0f1209dd6fa04ef02056ff"
+
+  depends_on "node"
+
+  def install
+    system "npm", "install", *Language::Node.local_npm_install_args
+    system "npm", "run-script", "build"
+    prefix.install ['dist', 'osascript']
```
Comment (ilovezfs) [inline Formula/imageoptim-cli.rb]:
```
libexec.install "dist", "osascript"
```

File: Formula/imageoptim-cli.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+require "language/node"
+
+class ImageoptimCli < Formula
+  desc "CLI interface for ImageOptim, ImageAlpha and JPEGmini"
+  homepage "https://jamiemason.github.io/ImageOptim-CLI/"
+  url "https://github.com/JamieMason/ImageOptim-CLI/archive/2.0.2.tar.gz"
+  sha256 "65ccf26cf45012e20877bf7985b1f9d99cfef4ac3d0f1209dd6fa04ef02056ff"
+
+  depends_on "node"
```
Comment (ilovezfs) [inline Formula/imageoptim-cli.rb]: Isn't this build time only?

Comment (reitermarkus) [reply Formula/imageoptim-cli.rb]: Yes.

Comment (ilovezfs) [reply Formula/imageoptim-cli.rb]: Then `  depends_on "node" => :build`

File: Formula/imageoptim-cli.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+require "language/node"
+
+class ImageoptimCli < Formula
+  desc "CLI interface for ImageOptim, ImageAlpha and JPEGmini"
```
Comment (ilovezfs) [inline Formula/imageoptim-cli.rb]: The I in CLI is interface

**Review by ilovezfs — APPROVED** [review]

**Comment by reitermarkus** [comment]
> Try downloading the bottle file and installing it locally.

How do I do that?

**Comment by ilovezfs** [comment]
https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/25678/version=high_sierra/artifact/imageoptim-cli-2.0.2.high_sierra.bottle.tar.gz

**Comment by reitermarkus** [comment]
The obvious difference is that the binary in the bottle has 37.3 MB, locally it has 38.2 MB. Any suspicion what could be wrong?

**Comment by ilovezfs** [comment]
My guess would be something having to do with the terminal type.

**Comment by ilovezfs** [comment]
Yup. `script -q /dev/null` somehow gets around the issue.

**Comment by reitermarkus** [comment]
So what does `script` do? Simulate a TTY?

**Comment by ilovezfs** [comment]
Yes, I think so. It's possible `PTY.spawn` would work here, too.

I'm wondering if this is an `npm` bug. CC @chrmoritz

**Comment by chrmoritz** [comment]
I think this is more likely caused by `nexe` itself instead of `npm`. It looks similar to the `nexe` issue https://github.com/nexe/nexe/issues/467 with a possible woraround in https://github.com/nexe/nexe/issues/467#issuecomment-388252022. Maybe patching the `package.json` to `"build:bin": "nexe --target 'mac-x64-10.0.0' --enableStdIn='' --input dist/imageoptim.js --output dist/imageoptim",` will work.

But do we really want to ship this as a js script bundled into a (possible duplicated) node executable? We could also just depend on `node` at runtime too and run the transpiled js script against our node executeable and skipping the whole `nexe` step.

**Comment by ilovezfs** [comment]
@chrmoritz what would that look like?

**Comment by chrmoritz** [comment]
@ilovezfs It would require to run the ts transpiler during the build, which is unfortunately not supported by the current version of language node, because https://github.com/Homebrew/brew/pull/2820 never got merged. But I plan to revive this PR based on the new `npm ci` command (for faster lock-file based installs) soon anyway.

For now this formula would require to duplicate some language node internals to work that way. 

Here is a completely working version of this formula based on a runtime node dep without nexe, which I've quickly hacked together (could still need some improvements): https://gist.github.com/chrmoritz/0a4f37d0b3e8f52e3243f3c0c89bf8a6

**Comment by ilovezfs** [comment]
I think unless imageoptim-cli upstream adds an install script with support for that, which would be cool, we're probably best off with the `nexe` approach to keep the formula as simple as possible.

**Comment by reitermarkus** [comment]
By the way, bug can be reproduced locally with `printf '' | brew install imageoptim-cli`.

**Comment by chrmoritz** [comment]
@ilovezfs I agree for now. But with proper language node support for this, getting it to work would be as simple as patching the `package.json` bin to the js file, renaming `build:ts` to `prepare`, running the improved language node on it and adding the missing shebang to the js file.

**Comment by reitermarkus** [comment]
Problematic code is probably this: https://github.com/nexe/nexe/blob/dev/src/options.ts#L202-L204

File: Formula/imageoptim-cli.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  homepage "https://jamiemason.github.io/ImageOptim-CLI/"
+  url "https://github.com/JamieMason/ImageOptim-CLI/archive/2.0.2.tar.gz"
+  sha256 "65ccf26cf45012e20877bf7985b1f9d99cfef4ac3d0f1209dd6fa04ef02056ff"
+
+  depends_on "node" => :build
+
+  def install
+    inreplace "package.json", /("build:bin": "nexe .*)(")/, "\\1 --enableStdIn ''\\2"
+
+    system "npm", "install", *Language::Node.local_npm_install_args
```
Comment (chrmoritz) [inline Formula/imageoptim-cli.rb]: You could use [`npm ci`](https://docs.npmjs.com/cli/ci#description) to get faster `package-lock.json` based installs here.

```ruby
system "npm", "ci", *Language::Node.local_npm_install_args
```

**Comment by chrmoritz** [comment]
@ilovezfs Here is a version of the formula without nexe and with runtime node using my wip language node branch https://github.com/chrmoritz/brew/commit/fd3073989453f7b3629d0cbf7b9fbd35856effc8: https://gist.github.com/chrmoritz/0a4f37d0b3e8f52e3243f3c0c89bf8a6

**Comment by ilovezfs** [comment]
@chrmoritz interesting. The question to me is whether you could get it to the point that it doesn't require any string replacements. That might require a little participation from upstream.

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
Thanks for the 🆕 formula, @reitermarkus! 🖼

**Pull request closed without merging**